### PR TITLE
Enable item collision detection for sudden movement

### LINKED ISSUE #6433: "physical" property of item entities
Entities have a property "[physical](https://github.com/minetest/minetest/blob/27144b471678b30156ad3383fb8d26a5bd9b66cb/doc/lua_api.txt#L4110)" which is barely documented and decides whether collision detection is performed for the entity.
The collision detection is done before on_step of the entity is called, see [content_sao.cpp](https://github.com/minetest/minetest/blob/27144b471678b30156ad3383fb8d26a5bd9b66cb/src/content_sao.cpp#L339-L394).

For item entities, this is a problem, because [item_entity.lua](https://github.com/minetest/minetest/blob/27144b471678b30156ad3383fb8d26a5bd9b66cb/builtin/game/item_entity.lua#L151) sets the physical property afterwards, so if the velocity is changed before the collision detection, as happens with the TNT mod from minetest_game, the physical property is false and the item can move inside nodes (this happens very consistently).
See also #5482.
This can be solved by executing the on_step functions before (could be problematic elsewhere), or by setting the physical property directly after the velocity change, but according to celeron55 requiring the mod to do so is not intended.

Always performing the collision detection does not change much for items, as it is not performed if the entity is not moving, but it needs to be disabled for carts ([cart_entity.lua](https://github.com/minetest/minetest_game/blob/4ccafd15187cf70d06d68ad3931912a3c3c4b4f5/mods/carts/cart_entity.lua#L2)), where physical=true is set.

Additionaly, the code in [item_entity.lua](https://github.com/minetest/minetest/blob/27144b471678b30156ad3383fb8d26a5bd9b66cb/builtin/game/item_entity.lua#L151) is incomplete, because items are stopped immediately if the node it is on is not defined (wanted that way), if it is walkable and not slippery or if the y-component of the velocity is zero, regardless of the x- and z-components.

I could not reproduce #5966, which could also be caused by only checking for the y direction.
#5973 and #5297 are unrelated, I think.

So it probably would be best to set physical=true for items unconditionally, because, as mentioned, [collisionMoveSimple](https://github.com/minetest/minetest/blob/27144b471678b30156ad3383fb8d26a5bd9b66cb/src/collision.cpp#L242-L244) already skips the detection if the entity is not moving, and to stop (or decelerate if on a slippery node) them like before.

### PR DESCRIPTION
**Pull request by DTA7**
Fixes #6433.

### CODE REVIEW

**Comment by paramat** [comment]
@SmallJoker

File: builtin/game/item_entity.lua [hunk 1]
```diff
@@ -148,49 +148,46 @@ core.register_entity(":__builtin:item", {
 		local vel = self.object:getvelocity()
 		local def = node and core.registered_nodes[node.name]
 		-- Ignore is nil -> stop until the block loaded
-		local is_physical = (def and not def.walkable) or vel.y ~= 0
+		local is_moving = (def and not def.walkable) or
+				(vel.x ~= 0 or vel.y ~= 0 or vel.z ~= 0)
```
Comment (paramat) [inline builtin/game/item_entity.lua]: Brackets not needed, might be faster without them

Comment (DTA7) [reply builtin/game/item_entity.lua]: Should I remove the brackets around "def and not def.walkable" as well?

Comment (paramat) [reply builtin/game/item_entity.lua]: No, because there they are needed to define the order of computation: AND NOT first, then OR OR OR.

Comment (SmallJoker) [reply builtin/game/item_entity.lua]: Ironically you were totally right about the first comment to remove \*brackets\* as the logical AND has a higher priority than OR. Definitions. JFYI.
```Lua
> return ({} and not true) or (0 ~= 0 or -5 ~= 0 or 0 ~= 0)
true
> return {} and not true or 0 ~= 0 or -5 ~= 0 or 0 ~= 0
true
> return ({} and not true) or (0 ~= 0 or 0 ~= 0 or 0 ~= 0)
false
> return {} and not true or 0 ~= 0 or 0 ~= 0 or 0 ~= 0
false
```

Comment (DTA7) [reply builtin/game/item_entity.lua]: The precedence of not is higher than the one of and, which is higher than the one of or, right?
But that would be confusing, I'll leave them.

Comment (paramat) [reply builtin/game/item_entity.lua]: Yeah, but AND having higher precedence than OR is not obvious, i forget that all the time, so yes it's there to make precedence obvious and clear, whereas the brackets around the 'vel's do not help with clarity at all and are actually confusing.

Comment (DS-Minetest) [reply builtin/game/item_entity.lua]: http://www.lua.org/manual/5.1/manual.html#2.5.6

**Comment by paramat** [comment]
:+1:

**Comment by paramat** [comment]
ce832eb4218340d92caa8961f5b7d7c2cda47a7e

**Pull request closed without merging**